### PR TITLE
Add Registry for registering rate limit algorithm engines

### PR DESCRIPTION
**Pull request by Shareed2k**
some benchmark with lru cache

```bash
➜  go_redis_ratelimit git:(fix_expr_time) ✗ go test -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/bringg/go_redis_ratelimit
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
Benchmark_CloudflareAlgorithm-12           17674             67348 ns/op            2161 B/op         89 allocs/op
Benchmark_GcraAlgorithm-12                  2280            524248 ns/op          332285 B/op       2401 allocs/op
Benchmark_SlidingWindowAlgorithm-12         2127            944436 ns/op          460514 B/op       2404 allocs/op
PASS
ok      github.com/bringg/go_redis_ratelimit    7.794s
```

with just map as cache
```bash
➜  go_redis_ratelimit git:(fix_expr_time) go test -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/bringg/go_redis_ratelimit
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
Benchmark_CloudflareAlgorithm-12           19647             66900 ns/op            2144 B/op         88 allocs/op
Benchmark_GcraAlgorithm-12                  2205            512690 ns/op          338248 B/op       2415 allocs/op
Benchmark_SlidingWindowAlgorithm-12         2254            901020 ns/op          461796 B/op       2401 allocs/op
PASS
ok      github.com/bringg/go_redis_ratelimit    6.473s
```

```bash
➜  go_redis_ratelimit git:(master) go test -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/bringg/go_redis_ratelimit
cpu: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
Benchmark_CloudflareAlgorithm-12         	   13971	     98175 ns/op	    2454 B/op	     100 allocs/op
Benchmark_GcraAlgorithm-12               	    2025	    553962 ns/op	  338609 B/op	    2415 allocs/op
Benchmark_SlidingWindowAlgorithm-12      	    2020	    968296 ns/op	  455437 B/op	    2405 allocs/op
Benchmark_SlidingWindowV2Algorithm-12    	    3949	    562815 ns/op	   59221 B/op	     162 allocs/op
PASS
ok  	github.com/bringg/go_redis_ratelimit	9.515s
```

### CODE REVIEW

File: algorithm/cloudflare/redis_store.go [hunk 1]
```diff
@@ -11,7 +11,8 @@ import (
 )
 
 type RedisDataStore struct {
-	RDB algorithm.Rediser
+	RDB            algorithm.Rediser
+	ExpirationTime time.Duration
```
Comment (shamil) [inline algorithm/cloudflare/redis_store.go:15]: Do we need it here? Cant' we have just `window * 2` in the `func (s *RedisDataStore) Inc(key string, window time.Time) error` function directly? or the `window` is not same as `period`?

File: algorithm/model.go [hunk 2]
```diff
@@ -2,34 +2,43 @@ package algorithm
...
 	Limit interface {
-		GetAlgorithm() string
-		GetBurst() int64
 		GetRate() int64
+		GetBurst() int64
+		GetAlgorithm() string
 		GetPeriod() time.Duration
 	}
 
 	Rediser interface {
```
Comment (shamil) [inline algorithm/model.go:23]: what happened with the sorting here?

File: limit.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import "time"
+
+type (
+	Limit struct {
+		Algorithm string
+		Burst     int64
+		Rate      int64
+		Period    time.Duration
+	}
+)
```
Comment (shamil) [inline limit.go:12]:
```suggestion
type Limit struct {
    Algorithm string
    Burst     int64
    Rate      int64
    Period    time.Duration
}
```

File: limiter.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
+package go_redis_ratelimit
+
+import (
+	"errors"
+
+	"github.com/go-redis/redis/v8"
+
+	"github.com/bringg/go_redis_ratelimit/algorithm"
+
```
Comment (shamil) [inline limiter.go:9]: remove blank line

File: limiter.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	_ "github.com/bringg/go_redis_ratelimit/algorithm/all"
+)
+
+const (
+	DefaultPrefix = "limiter"
+)
+
+var (
+	algorithmPool map[string]algorithm.Algorithm
```
Comment (shamil) [inline limiter.go:18]: any reason it's global var and not part of the `Limiter` struct?

File: limiter.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/go-redis/redis/v8"
+
+	"github.com/bringg/go_redis_ratelimit/algorithm"
+
+	_ "github.com/bringg/go_redis_ratelimit/algorithm/all"
+)
+
+const (
+	DefaultPrefix = "limiter"
+)
```
Comment (shamil) [inline limiter.go:15]: just a nit, but having this one line const, makes code look cleaner, unless you have multiple const definitions
same for other things, like `var` and `type`

File: algorithm/model.go [hunk 7]
```diff
@@ -60,4 +69,23 @@ type (
...
+		Name         string
+		NewAlgorithm func(rdb Rediser) (Algorithm, error)
+	}
 )
+
+func Register(info *RegInfo) {
+	Registry = append(Registry, info)
+}
+
+func Find(name string) (*RegInfo, error) {
```
Comment (shamil) [inline algorithm/model.go:83]: Where we use this method?

File: limiter.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+type (
+	// Limiter controls how frequently events are allowed to happen.
+	Limiter struct {
+		Prefix string
+	}
+)
+
+// NewLimiter returns a new Limiter.
+func NewLimiter(rdb *redis.Client) (*Limiter, error) {
+	algorithmPool = make(map[string]algorithm.Algorithm, len(algorithm.Registry))
```
Comment (shamil) [inline limiter.go:30]:
```suggestion
	algorithms = make(map[string]algorithm.Algorithm, len(algorithm.Registry))
```

File: algorithm/all/all.go [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package all
```
Comment (shamil) [inline algorithm/all/all.go:1]: maybe move this file to  `algorithm/all.go`?

**Review by shamil — CHANGES_REQUESTED** [review]

File: algorithm/cloudflare/redis_store.go (hunk 1 above)
Comment (Shareed2k) [reply algorithm/cloudflare/redis_store.go:15]: yes this is not a same, can't do it inside `Inc` function

File: algorithm/model.go (hunk 2 above)
Comment (Shareed2k) [reply algorithm/model.go:23]: i love how it looks from short to long ;)

File: algorithm/all/all.go (hunk 9 above)
Comment (Shareed2k) [reply algorithm/all/all.go:1]: can't do it, will be cyclic import

**Review by shamil — APPROVED** [review]

**Pull request merged**